org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPEclipseUtils.java

```diff
@@ -562,27 +563,46 @@ public static IDocument getDocument(@Nullable IResource resource) {
 		if (resource == null) {
 			return null;
 		}
+		IDocument document = null;
+		synchronized(connectedFiles) {
+			document = getExistingDocument(resource);
 
-		IDocument document = getExistingDocument(resource);
+			if (document == null && resource.getType() == IResource.FILE) {
+				ITextFileBufferManager bufferManager = FileBuffers.getTextFileBufferManager();
+				if (bufferManager == null)
+					return document;
+				try {
+					bufferManager.connect(resource.getFullPath(), LocationKind.IFILE, new NullProgressMonitor());
+					connectedFiles.add(resource.getFullPath());
+				} catch (CoreException e) {
+					LanguageServerPlugin.logError(e);
+					return document;
+				}
 
-		if (document == null && resource.getType() == IResource.FILE) {
-			ITextFileBufferManager bufferManager = FileBuffers.getTextFileBufferManager();
-			if (bufferManager == null)
-				return document;
-			try {
-				bufferManager.connect(resource.getFullPath(), LocationKind.IFILE, new NullProgressMonitor());
-			} catch (CoreException e) {
-				LanguageServerPlugin.logError(e);
-				return document;
+				ITextFileBuffer buffer = bufferManager.getTextFileBuffer(resource.getFullPath(), LocationKind.IFILE);
+				if (buffer != null) {
+					document = buffer.getDocument();
+				}
 			}
+		}
+		return document;
+	}
 
-			ITextFileBuffer buffer = bufferManager.getTextFileBuffer(resource.getFullPath(), LocationKind.IFILE);
-			if (buffer != null) {
-				document = buffer.getDocument();
+	public static void disconnectFromFileBuffer(@Nullable IPath path) {
+		synchronized(connectedFiles) {
+			if (path == null || !connectedFiles.contains(path)) {
```

Comment:
could we use just `remove` and use the returned `boolean` to exit early instead of calling `contains` and then `remove`?